wsdl-core/src/main/java/io/ballerina/wsdl/core/WSDLToBallerina.java

```diff
@@ -0,0 +1,240 @@
+            reader.setFeature("javax.wsdl.importDocuments", true);
+
+            InputStream wsdlStream = new ByteArrayInputStream(wsdlText.getBytes(Charset.defaultCharset()));
+            wsdlDefinition = reader.readWSDL(null, new InputSource(wsdlStream));
+
+            WSDLService wsdlService = getSOAPService(wsdlDefinition);
+            if (wsdlService == null) {
+                return response;
+            }
+
+            soapVersion = wsdlService.getSoapVersion();
+            initializeSchemas(wsdlDefinition);
+
+            List<WSDLOperation> wsdlOperations = getWSDLOperations();
+            wsdlService.setWSDLOperations(wsdlOperations);
+
+            TypeGenerator typeGenerator = new TypeGenerator(wsdlService, soapVersion);
+
+            try {
+                String typeModulePart = typeGenerator.generate(filteredWSDLOperations);
+                GeneratedSourceFile typesSrc = new GeneratedSourceFile("types.bal", typeModulePart);
+                response.setTypesSource(typesSrc);
+            } catch (FormatterException e) {
+                DiagnosticMessage message = DiagnosticMessage.wsdlToBallerinaErr101(null);
+                diagnosticMessages.add(message);
+            }
+        } catch (WSDLException e) {
+            DiagnosticMessage message = DiagnosticMessage.wsdlToBallerinaErr100(null);
+            diagnosticMessages.add(message);
+        }
+        return DiagnosticUtils.getDiagnosticResponse(diagnosticMessages, response);
+    }
+
+    private void initializeSchemas(Definition wsdlDefinition) {
+        Map<String, XmlSchema> targetNSToSchema = new HashMap<>();
+        Types types = wsdlDefinition.getTypes();
+        List<?> extensions = new ArrayList<>();
+        if (types != null) {
+            extensions = wsdlDefinition.getTypes().getExtensibilityElements();
+        }
+        for (Object extension : extensions) {
+            if (extension instanceof Schema) {
+                Element schElement = ((Schema) extension).getElement();
+                String targetNamespace = schElement.getAttribute("targetNamespace");
+                if (!targetNamespace.isEmpty()) {
+                    targetNSToSchema.put(targetNamespace, new XmlSchemaCollection().read(schElement));
+                }
+            }
+        }
+        SchemaManager.getInstance().initializeSchemas(targetNSToSchema);
+    }
+
+    private WSDLService getSOAPService(Definition wsdlDefinition) {
+        @SuppressWarnings("unchecked")
+        Collection<Service> services = wsdlDefinition.getAllServices().values();
+        for (Service service : services) {
+            @SuppressWarnings("unchecked")
+            Collection<Port> ports = service.getPorts().values();
+            for (Port port : ports) {
+                List<?> extensions = port.getExtensibilityElements();
+                for (Object extension : extensions) {
+                    if (extension instanceof SOAPAddress || extension instanceof SOAP12Address) {
```

Comment:
I would rather decouple this condition into if and if-else block. That would simplify the mental mapping and improve the readability. And it would allow us to use Java instanceOf pattern-matching [1] and remove the redundant casting.

[1] - https://www.baeldung.com/java-pattern-matching-instanceof